src/librustc_resolve/lib.rs

```diff
@@ -2505,29 +2489,23 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         self.ribs[TypeNS].pop();
     }
 
-    fn with_tuple_struct_self_ctor_rib<F>(&mut self, self_ty: &Ty, f: F)
+    fn with_tuple_struct_self_ctor_rib<F>(&mut self, self_ty: &Ty, impl_id: DefId, f: F)
         where F: FnOnce(&mut Resolver)
     {
-        let variant_def = if self.session.features_untracked().tuple_struct_self_ctor {
-            let base_def = self.def_map.get(&self_ty.id).map(|r| r.base_def());
-            if let Some(Def::Struct(ref def_id)) = base_def {
-                self.tuple_structs.get(def_id).cloned()
-            } else {
-                None
+        let base_def = self.def_map.get(&self_ty.id).map(|r| r.base_def());
+        match base_def {
+            Some(Def::Struct(..)) |
+            Some(Def::TyAlias(..)) => {
```

Comment:
This check is unnecessary, you can always add the self constructor rib.
Whether the type is a struct or not needs to be checked in typeck anyway, and you'll get better diagnostics if `Self` referring to e.g. enums is reported there.

Reply:
Fixed.